### ISSUE TITLE
Language initiative (for 2.0.0?)

### ISSUE THREAD

**Issue by Gormartsen** [backdrop/backdrop-issues#1913]
Me and @MrHaroldA had a talk today about better multilingual support in backdrop.

Right now if you want to localize backdrop you need to import drupal 7 translation file.

For for 2.0.0 we do have in roadmap :
>  ...  this likely includes expanding language support

I propose to setup our own https://localize.*** website but instead of coping drupal.org localize process, do it different way:

- use RESTful JSON based API server (no xmlrpc)
- provide voting process instead of having admin to approve translation
- load translation from server right in a moment of activating new language (optional checkmark?)
- push translation right from translation page of user website, when he translate some string (optional checkmark: share translation) aka l10_client

I personally out of hours per day but I can do next input:
 - provide infrastructure (server where  https://localize.*** website lives)
 - mentor process of writing API server without using monster modules aka http://drupal.org/project/services
 - mentor process of writing a JSON translate client based on [backdrop_http_request](https://api.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_http_request/1)
 - testing

**Comment by quicksketch** [comment]
This is a great idea. Multilingual is one of the major areas Backdrop has trouble when compared to Drupal 8, because Drupal 7 really wasn't that great at multilingual either.

From the description here, could we possibly downgrade this issue from "Multilingual initiative" which is very broad, to just "Set up a Localization Server and Integrate it into Backdrop?" (which still is pretty broad).

I don't think there's any reason why we would need to wait for 2.0 in order to do the things mentioned in the summary, if we can implement this as feature additions and without API changes.

> push translation right from translation page of user website, when he translate some string (optional checkmark: share translation) aka l10_client

There are some challenges here, but I think they could be overcome. When strings are run through `t()`, Backdrop doesn't have any idea what module it's coming from. So contributing strings for particular modules could prove tricky. Lots of strings are going to be site-specific and wouldn't have any value to the central server. And different modules may translate the same string repeatedly. But when you turn on a module for the first time, Backdrop will need to be able to ask the translation server, "Give me all the strings for X module, version Y." so there has to be some association between the strings and the modules that require them.

So the translation server must know what strings are in each version of a module. Drupal uses the https://www.drupal.org/project/potx module for this purpose, finding all strings wrapped in `t()`, `st()`, and special places like `'title'` items in `hook_menu()`.

If the translation server knew what strings were in which modules, then the contributing strings process would look like this:

- Contributing site: "Hi localize.backdropcms.org. I have these 45 modules enabled. What strings are needed for these 45 modules?"
- localize.backdropcms.org: "Based on your 45 modules, I know what strings are in 20 of them. Please provide translations for these 600 strings associated with these 20 modules."
- Contributing site: "I have 240 translations for the requested 600 strings. Here you go."
- localize.backdropcms.org: "Thank you contributor. New translations are packaged every X hours. You may review/vote on contributed strings at [url]."

This arrangement makes it so that the potx module for extracting strings only needs to live one the central server. And when new modules are installed, the translation server can be queried for the strings associated with that module/version.

**Comment by klonos** [comment]
> From the description here, could we possibly downgrade this issue from "Multilingual initiative" which is very broad, to just "Set up a Localization Server and Integrate it into Backdrop?" (which still is pretty broad).

Yes, we could do that. Or make this a meta. Anyways, baby-steps++. Lets draft a plan and set tasks (like the localization server) and tackle them one at a time.

> So the translation server must know what strings are in each version of a module. Drupal uses the https://www.drupal.org/project/potx module for this purpose, finding all strings wrapped in `t()`, `st()`, and special places like `'title'` items in `hook_menu()`.

> If the translation server knew what strings were in which modules, ...

How about this then... we have potx.module (or whatever we choose to call it) make a list of all the strings and also *for each string* have it keep a record of the following details:

- Core module (or theme because these have settings too) or contrib. We can do this by having the module first check module and theme directories under `/core/*` and then move to checking the respective directories in the root folder. So we have core vs contrib.
- Project name (we can extract this from the path)
- Project version from the respective .info file If core, then core version from `'BACKDROP_VERSION'`.
- Project type? (not sure we need this though, but we might in the future)

**Comment by quicksketch** [comment]
> make a list of all the strings and also for each string have it keep a record of the following details:

Yep, all that sounds good. On BackdropCMS.org we just store the "project_name" as "backdrop" for core, and the "project_type" becomes "core" instead of "module", "theme", or "layout". I think if this were in MySQL you'd end up with a table of strings, each a string ID. Then a table of translations by string ID => translation. And a 3rd table of all meta data, e.g.

string:

| string_id | string |
|-----------|----------|
| 1         | Hello world!

string_translation:

| string_id | langcode | translation |
|-----------|----------|-------------|
| 1         | fr | Bonjour monde!

string_usage:

| string_id | project_name | project_type | version   |
|-----------|--------------|--------------|-----------|
| 1         | my_module    | module       | 1.x-1.0.0 |

Anyway, for basic storage/retrieval/interface for this, it seems like we should be able to take most of localize.drupal.org and port it, no? The API part we can write however we like, but it might be wasted effort to write this from the ground-up. We should see what parts we can easily leverage and port them.

**Comment by Gormartsen** [comment]
> Multilingual initiative" which is very broad, to just "Set up a Localization Server and Integrate it into Backdrop?"

👍 

> When strings are run through t(), Backdrop doesn't have any idea what module it's coming from. So contributing strings for particular modules could prove tricky.

All strings are swappable. t('Delete') will be used for all modules include contrib. 
There is no way to have Delete translated one way for one module and another for another.

From this perspective I don't see need to provide details where this translation coming from, when user post a translation. Only language that it's relate to.

On localize.backdropcms.org we can have a parser, that get data from modules and themes and core and associate them to specified module, theme or core.

@quicksketch ++ to basic table structure.

I would add extra - alternative:

| string_id | langcode | translation | votes
|-----------|----------|-------------|------|
| 1         | ua |  Витаемо мир! | 10000
| 1         | ua | Привіт Світ! | 9999

> Anyway, for basic storage/retrieval/interface for this, it seems like we should be able to take most of localize.drupal.org and port it, no? The API part we can write however we like, but it might be wasted effort to write this from the ground-up. We should see what parts we can easily leverage and port them.


I would say that UX is horrible on localize.drupal.org. 

I prefer to start with a question 'what is end user needs?'
- imported translation as soon as he selected new language on install backdropcms or on language settings page
- share his changes without going nuts (simply one button 'share this')
- if he feel so, add his vote to translation . User case: after import, user does not like current TOP translation for button Hello world!. He changes it to another one. We can even suggest few one based on top votes. By selecting it, user add his vote to it. (Some security protection required but not on this stage agains trolls)

That's it. 
There is advanced user who are focused to do more translations just because he can.
His needs I see next way:
- see what is not translated and add his translation (+1 vote)
- see what is translated with alternatives and vote or add translation
- see new strings that added recently and are waiting for translation

Advanced user is main user for  localize.drupal.org. Regular user should go there only if he is interested to become advanced. Everything else admin part of his site should take care.

There is a need to do more research on advanced user needs.

**Comment by Gormartsen** [comment]
Just for the record:
Need to check current translation changes compare to d7.
There is a performance boost idea of using Json file with translation strings to load on hook_init level.
It should provide performance boost specially on multilingual website with too many strings to display.

Right now in d7 to speed it up, you need to put all translation strings to settings.php.
Json file sounds better. It can be easily deployed to production without rebuild.

**Comment by olafgrabienski** [comment]
> could we possibly downgrade this issue from "Multilingual initiative" which is very broad, to just "Set up a Localization Server and Integrate it into Backdrop?"

> Yes, we could do that. Or make this a meta-issue. (...) Lets draft a plan and set tasks (like the localization server) and tackle them one at a time.

Still a good idea in my opinion! So, what's better: meta-issue or downgrade / rename the issue? I'm however interested in the topic because I'd like to help (once there is a infrastructure) with contributing to the German localization of Backdrop CMS.

**Comment by klonos** [comment]
And I can do the same with Greek with the help of @dyrer
